src/xdsl/ir.py

```diff
@@ -368,18 +368,35 @@ def verify(self, verify_nested_ops: bool = True) -> None:
     def verify_(self) -> None:
         pass
 
-    def clone_without_regions(self: OperationType) -> OperationType:
+    def clone_without_regions(self: OperationType,
+                              mapper=None) -> OperationType:
```

Comment:
Can you add the type of mapper here? At list `Dict`.

Also, should we maybe separate the operands_mapper to the successor mapper (I'm not entirely sure of this though).

Reply:
The use case where this is required is cloning a region with lots of operations, which also refer to another and possibly again have nested regions. 
consider the following example:
```
module() {
  %0 : !i32 = arith.constant() ["value" = 1 : !i32]
  %1 : !i32 = arith.constant() ["value" = 2 : !i32]
  %2 : !i32 = arith.addi(%0 : !i32, %1 : !i32)
  std.return(%2 : !i32)
}
```
the call to `module.clone()` passes a mapper to all the operations in the region, which the two constants use to register the mapping of their old result to the result of the cloned operations.
addi finds the mapping of results for the two constants and uses the updated values of the mapper as operands. 

So it is necessary for an operation being cloned in context of a region to add its results to the map and to look up possible changes in the operands in the map. Same goes the successors of an op. The mapping of old Block to new Block is added by the Region being cloned. 

We could divide the `mapper` into a `valueMapper` for operands and results and a `blockMapper` for successors if you like that better. 

Another thing: If this function does not get passed a mapper we can assume the op is being cloned in isolation. So this could also look like:
```
    def clone_without_regions(self: OperationType,
                              mapper=None) -> OperationType:
        if mapper is None:
            # branch old version
            operands = self.operands
            result_types = [res.typ for res in self.results]
            attributes = self.attributes.copy()
            successors = self.successors.copy()
            regions = [Region() for _ in self.regions]
            return self.create(operands=operands,
                           result_types=result_types,
                           attributes=attributes,
                           successors=successors,
                           regions=regions)
        else:
            # branch PR version, what is now starting at line 376
```

but I did not like the duplication in it. Maybe you have an idea for something more elegant?

Reply:
Done :+1: